### ISSUE TITLE
Issues setting $PATH (using read or set with variables)

### ISSUE THREAD

**Issue by simotek** [fish-shell/fish-shell#2968]
From the world of people trying to do crazy things on the internet...

As some background I am trying to create 1 or many ".env" files that fish will look at on startup, this for the most part works well unless you try to set the $PATH variable.

Broken Way 1 `echo $PATH | read -x PATH` Its best not to put that in your config.fish atm. But you can execute it in a terminal to get a very broken fish.

Broken Way 2 (seems to have different effects if done from the commandline vs config.fish
` set -x TMP '$PATH'`
`echo $TMP`
`set -x PATH $TMP`

Basically I have loaded a string containing `$PATH` into a variable and I'd like to use that to set the PATH.

---
Fish version: 2.3b1

Operating system: openSUSE Leap 42.1 (Beta repo)

Terminal or terminal emulator: Terminology

**Comment by krader1961** [comment]
Your first example converts an array of values into a single value with the original values space separated. This will work:

```
echo $PATH | read -a -x PATH
```

Your second example won't work because variables aren't interpolated inside single-quoted strings. But even if you switch to double-quotes you'll have the same problem as your first example: you're coercing an array of values to be a single value of space separate tokens.

Try this:

```
count $PATH
set -x TMP "$PATH"
count $TMP
```

Notice the difference? If you then try to run `set -x PATH $TMP` it should give you a warning and not change PATH.

PATH is a special-case in fish. When fish starts it reads the value of PATH from the environment and splits it on colons to produce an array of directories. When it runs an external command it does the reverse when putting it in the environment for that external command.

**Comment by simotek** [comment]
Hi, I know that `echo $PATH | read -a -x PATH` would work, I was trying to show the simplest way to break it (It is broken here, it should atleast have the same behavior as `set PATH` where it errors). The other command also fails in the same way which results in no path set (atleast my prompt cant find the command hostname).  I did see the other error you were mentioning before.

What I am trying to achieve is the following I have the following variable `FileLine` it contains the following `$PATH /home/simon/.local/bin` I'd like to expand $PATH then assign the result to PATH, I can probably do this using the new string substitution API but I'd prefer not to have to special case it.

Its ok if fish decides to have special case handling for $PATH but fish should prevent users from doing things that lead to fish becoming totally unusable :-) The above too examples both currently do that.

**Comment by krader1961** [comment]
> ... but fish should prevent users from doing things that lead to fish becoming totally unusable

That is an unrealistic goal. Should fish stop you from running `rm -rf /`? That will also make fish totally unusable.

If you have a variable that contains a directory you want to add to PATH just do

```
set -x PATH $PATH $FileLine
```

That will also do the right thing if FileLine is an array of directories.

**Comment by simotek** [comment]
well users expect `rm -rf /` to remove everything (Atleast on my distro  i'm almost certain it gives you a option to prevent it). The difference is the standard normal user who hasn't read the documentation would expect both of the commands listed above to work rather then leave them with a empty path.

**Comment by krader1961** [comment]
> ... the commands listed above to work rather then leave them with a empty path.

Those commands don't leave the PATH empty. They result in PATH containing a very long value that doesn't refer to a valid directory. For example, running your first example on my system results in this:

```
$ count $PATH
1
$ printf '|%s|\n' $PATH
|/usr/local/opt/coreutils/libexec/gnubin /usr/local/opt/findutils/bin /Users/krader/bin /Users/krader/sbin /Users/krader/symlinks /usr/local/sbin /usr/local/bin /usr/bin /bin /usr/sbin /sbin /opt/X11/bin /usr/local/go/bin|
```

I've omitted all the error messages. The important thing to notice is that PATH now contains a single value. If you execute that `printf` before the assignment you'll see something very different. On my system:

```
$ printf '|%s|\n' $PATH
|/usr/local/opt/coreutils/libexec/gnubin|
|/usr/local/opt/findutils/bin|
|/opt/X11/bin|
|/Users/krader/bin|
|/Users/krader/sbin|
|/Users/krader/symlinks|
|/usr/local/bin|
|/usr/local/go/bin|
|/usr/local/sbin|
|/usr/bin|
|/usr/sbin|
|/bin|
|/sbin|
```

> The difference is the standard normal user who hasn't read the documentation would expect both of the commands listed above to work rather then leave them with a empty path.

Maybe, but that is a big mistake. It's like expecting to be able the python interpreter to correctly execute bash statements or vice versa. Fish is explicitly not a 100% POSIX compliant shell.

**Comment by simotek** [comment]
The problem is the variable should contain $PATH not only making it consistent with all others but often I may have it set to `/opt/QtCreator/bin $PATH /home/simon/.local/bin` where the order is quite important. I could do a bunch of string manipulation and set PATH twice once with the arguments before and again with the arguments after but there must be a better way

**Comment by krader1961** [comment]
I can't make any sense of your last comment, @simotek. Nonetheless, it appears you don't understand that fish treats $PATH as an array of directories rather than a simple string of colon delimited values like most other shells (e.g., bash). Run the `printf` in my previous comment before and after you modify PATH. Also run `count $PATH` before and after.

**Comment by simotek** [comment]
If thats the case It would be nice if I could pipe an array to read

`string replace '$PATH' $PATH $TMP | string split ' ' | read -gx PATH` and have it work, 

`set -x TMP '$PATH /home/simon/.local/bin /opt /home/simon/src/Scripts'
`string replace '$PATH' $PATH $TMP | string split ' ' | read -x TMP2`
`echo $TMP2`
`/usr/bin`

**Comment by simotek** [comment]
@krader1961 I now understand that fish treats $PATH as a array, i'm just trying to figure out how to actually set it in a sane way. It needs to be possible to prepend and append user paths to the system PATH if its not you break a huge number of custom UNIX systems. I am trying to take a standard configuration file thats used by many applications for setting up environments and load them into fish, yes i'm expecting I'll have to do some translation but atm the amount I have to do is alot more then I was expecting.

**Comment by krader1961** [comment]
You keep using single-quotes where doing so makes no sense. Run

```
echo 'WTF1 $PATH'
echo "WTF2 $PATH"
```

See the difference?

**Comment by simotek** [comment]
YES

**Comment by simotek** [comment]
`# More expansive example`
`for line in (cat $argv[1])`
`  set -l split (string split = $line)`
`  printf "%s" (eval "echo $split[2]") | read -l $split[1]`
`end`

I left out a bunch of error checking and ignoring comments but m aybe you'll get the picture $argv[1] is a file, one line in that file is `PATH=/opt/QtCreator $PATH /home/simon/.local/bin` now how do I set the contents of $split[2] as $PATH expanding the $PATH as I go

**Comment by simotek** [comment]
Ok well the following seems to work

`set -x TMP '$PATH /home/simon/.local/bin /opt /home/simon/src/Scripts'`
`string replace '$PATH' $PATH $TMP | string join ' ' | read -x TMP2`
`set -x PATH (string split ' ' $TMP2)`

I'm open to improvements, i'm going to close this bug as it has far to much noise and create a new one specifically stating that `echo "FOO" | read -x PATH` and `set -x PATH FOO` should have the same protection as `set -x PATH "$PATH FOO"` because if were happy to have warnings in some cases we should have them in all.

**Comment by simotek** [comment]
Well my above solution ends up prepending everything behind /usr/bin rather then putting some in front of it, I give up this PATH stuff is just overly complex and STUPD

**Comment by krader1961** [comment]
> often I may have it set to /opt/QtCreator/bin $PATH /home/simon/.local/bin where the order is quite important.

That's trivial:

```
set -x PATH /opt/QtCreator/bin $PATH /home/simon/.local/bin
```

So presumably you're trying to do something else which you finally [mention](https://github.com/fish-shell/fish-shell/issues/2968#issuecomment-213892141):

> ... one line in that file is `PATH=/opt/QtCreator $PATH /home/simon/.local/bin` now how do I set the contents of $split[2] as $PATH expanding the $PATH as I go.

Are you trying to do this in a manner that will work in fish and other shells like bash? If you're not then you're making this more complicated than it needs to be. Also, why are you exporting TMP and TMP2 in your examples?

> I give up this PATH stuff is just overly complex and STUPD

I appreciate that you're frustrated but manipulating the PATH is not overly complex. In fact, in fish it's much easier than in bash. In your "env" files why not just put valid fish statements then `source` the file? For example:

```
set -x PATH  /opt/QtCreator $PATH /home/simon/.local/bin
set -x PATH $PATH /another/directory/to/be/appended
set -x PATH /prepend/this/dir $PATH
```

**Comment by simotek** [comment]
Yeah sorry that was my fault, got bitten by #2971 and I was only exporting the TMP vars out of lazyness.

Yes what i've done is more complicated here, but the result is simpler more universal more shareable config files I put the code below if your interested, thanks for your time and help.

    function __tacklebox_load_env_file --no-scope-shadowing --description \
        'Load and export all values in a env file'
    # Loop through lines of .env file
    for line in (cat $argv[1])
        # If line starts with # presume it is empty or if its empty ignore it
        if test (string sub -s 1 -l 1 $line) != "#" -a -n $line
            # Check that string contains either exactly 1 = or exactly 1 ' '
            if string match -q "*=*" $line
                if not string match -q "*=*=*" $line
                    set -l split (string split = $line)
                    # Using the following does not work as it does not substitute the path
                    # Setting the PATH with read breaks
                    if test $split[1] != PATH
                        # need to expand $split[2] twice so that any vars stored in the file get expanded
                        set -l TMP "echo $split[2]"
                        printf "%s" (eval $TMP) | read -x $split[1]
                    else
                        # Fish handles PATH specially and must be handled specially
                        # Handle : or ' ' seperation of paths as thats what people expect
                        string replace '$PATH' (string join ' ' $PATH) $split[2] | string join ' ' | string replace -a ':' ' ' | read -l TMP
                        set -x PATH (string split ' ' $TMP)
                    end
                else
                    echo "Invalid line not added to environment: $line"
                end
            else
                echo "Invalid line not added to environment: $line"
            end
        end
       end
    end

    function __tacklebox_unload_env_file --no-scope-shadowing --description \
        'Clears all environment variables loaded in a env file'
        # Loop through lines of .env file
        for line in (cat $argv[1])
            # If line starts with # presume it is empty or if its empty ignore it
            if test (string sub -s 1 -l 1 $line) != "#" -a -n $line
                # Check that string contains either exactly 1 = or exactly 1 ' '
                if string match -q "*=*" $line
                    if not string match -q "*=*=*" $line
                        set -l split (string split = $line)
                        set -ex $split[1]
                    else
                        # echo "Invalid line not added to environment: $line"
                    end
                else if string match -q "* *" $line
                    if not string match -q "* * *" $line
                        set -l split (string split ' ' $line)
                        set -ex $split[1]
                    else
                        echo "Invalid line not added to environment: $line"
                    end
                else
                    echo "Invalid line not added to environment: $line"
                end
            end
        end
    end

But that makes the environment variable file which are the ones that get edited all the time rather then once really simple like so

    # Default Environment
    EDITOR=vim
    PATH=/opt/spotify/usr/bin/ $PATH /home/simon/.local/bin /home/simon/src/Scripts
    # Fun test Enlightenment exports the variable PANTS as ON
    SHIRT=$PANTS

**Comment by faho** [comment]
>`# Loop through lines of .env file
    for line in (cat $argv[1])`

This really calls for a `while read` loop with a redirection (`while read VAR; DOSTUFF; end < FILE`). Needs a `test -r` before, though.

>` # If line starts with # presume it is empty or if its empty ignore it
        if test (string sub -s 1 -l 1 $line) != "#" -a -n $line`

What you want is `string match -v` (added a short time ago). I.e. `string replace -r '#.*' '' | string match -rv '^\s*$'` - this will strip all comments (even in the same line) and remove empty lines.

> # Check that string contains either exactly 1 = or exactly 1 ' '

Are you defining the file format here or should this read arbitrary existing .env files? If you are, then why not define just one delimiter? If you aren't, you might want to look at how I dealt with this in share/functions/abbr.fish. Here's a few pointers:

- You should allow a delimiter to appear later on, but only use the first occurence to split (`string split -m 1`)

- You need to distinguish between delimiters by seeing which occurs _first_ (`string match -qr '^[^ ]+='` will be true if you have a '=' before any space)

>`set -ex $split[1]`

I'm assuming you know this, but this would be _really_ bad for the $PATH case (and a little less bad but still bad for $EDITOR), as it would _delete the entire thing_. You'd need to save the old state somewhere (e.g. make a `$PATH_old` or something, or create a "$OLDENV" that contains `VAR=VALUE` - you can then get the index by doing `contains -i PATH (string replace -r '=.*' '' -- $OLDENV)`).

**Comment by simotek** [comment]
Thanks for the feedback @faho i'll look through and implement this tomorrow. As a interesting reference `set -e PATH` just restores you to using the system path, in the shell at least it would break other apps that use the PATH variable though either way `set -x PATH FOO` breaks it far worse see #2969 . The unload isn't used nearly as often as the load in fact by default i'm not using it. Its more designed so that you can have a .env file per say development environment and unload one before loading another. 

I had already decided to drop the space delimiter to make things simpler, i just hadn't got to it when I copy / pasted that. 

Thanks for the other pointers i'll incorporate them
